guides/source/active_record_querying.md

```diff
@@ -1599,6 +1599,83 @@ WHERE reviews.id IS NULL
 
 Which means "return all customers that have not made any reviews".
 
+Using Common Table Expressions (CTE)
+-----------------------------
+
+[Common Table
+Expressions](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#Common_table_expression)
+are named temporary query results that are referable within a single SQL query.
+The [`with`][] method is a convenient way to access CTEs `WITH` syntax.
+
+NOTE: To use the CTEs on MySQL, MySQL version 8.0+ is required. On SQLite,
+version 3.8.3+ is required.
+
+The `with` method returns an instance of `ActiveRecord::Relation`, so you can
+chain other query methods and even refer to the CTE from them (e.g. `from`,
+`select`, `joins` or `left_outer_joins`):
+
+```ruby
+relation = Book
+  .with(reviewed_books: Review.select(:book_id).distinct)
+  .left_outer_joins(:reviewed_books)
+  .select("books.*, reviewed_books.book_id AS has_reviews")
+```
+
+The above should generate:
+
+```sql
+WITH "reviewed_books" AS (
+  SELECT DISTINCT "reviews"."book_id"
+  FROM "reviews"
+)
+SELECT books.*, reviewed_books.book_id AS has_reviews
+FROM "books"
+LEFT OUTER JOIN "reviewed_books"
+  ON "reviewed_books"."book_id" = "books"."id"
+```
+
+You can also define multiple CTEs by passing multiple key-value pairs:
+
+```ruby
+Book.with(
+  books_with_reviews: Book.where("reviews_count > ?", 0),
+  books_in_print: Book.in_print
+)
+```
+
+Or chaining multiple `with` calls:
+
+```ruby
+Book
+  .with(books_with_reviews: Book.where("reviews_count > ?", 0))
+  .with(books_in_print: Book.in_print)
+```
+
+These examples should generate:
+
+```sql
+WITH "books_with_reviews" AS (
+  SELECT "books".*
+  FROM "books"
+  WHERE (reviews_count > 0)
+),
+"books_in_print" AS (
+  SELECT "books".*
+  FROM "books"
+  WHERE "books"."in_print" = ?
+)
+SELECT "books".*
+FROM "books"
+```
+
+It is recommended to pass a query as `ActiveRecord::Relation`.
```

Comment:
Could you add an explanation about why it is recommended?
Is it because of composability?

Reply:
I assume he means the warning below about passing non-relation objects which might be susceptible to injection.

Reply:
@p8 @zzak  The sentence comes from https://github.com/rails/rails/blob/main/activerecord/lib/active_record/relation/query_methods.rb#L472C13-L472C24 , added in https://github.com/rails/rails/pull/37944 .
It seems that the specific reason of the recommendation cannot be found in https://github.com/rails/rails/pull/37944 .
The author of the PR might has an answer.

Reply:
Yeah, the reason is to just not pass untrusted input to avoid an injection attack. Maybe we can include more context there with the warning and join them to one concise thought.

Reply:
Added the reason for recommendation.